doc/man7/openssl-threads.pod

```diff
@@ -0,0 +1,98 @@
+Exceptions to this should be documented on the specific manual pages, and
+some general high-level guidance is given here.
+
+One major use of the OpenSSL thread API is to implement reference counting.
+Many object within OpenSSL are reference-counted, so resources are not
+released, until the last reference is removed.
+References are often increased automatically (such as when an B<X509>
+certificate object is added into an B<X509_STORE> trust store).
+There is often an I<object>_up_ref() API that can be used to increase the
+reference.
+Failure to match I<object>_up_ref() calls with the right number of
+I<object>_free() calls is a common source of memory leaks when a program
+exits.
+
+The various C<set1> and C<get1> functions associated with an object are
+used to set or get attributes of an object; they increase the refernece
+count.
+The C<set0> and C<get0> functions are similar, but do not increase the
+reference count, passing "ownership" appropriately.
+Functions with a simple C<set> or C<get> are historical; and the documentation
+must state how the references are handled.
+Get methods are often thread-safe as long as the ownership requirements are
+met.
+Set methods are generally not thread-safe as discussed below.
+
+Objects are thread-safe
+as long as the API's being invoked don't modify the object; in this
+case the parameter is usually marked in the API as C<const>.
+Not all parameters are marked this way.
+Note that a C<const> declaration does not mean immutable; for example
+L<X509_cmp(3)> takes pointers to C<const> objects, but the implementation
+uses a C cast to remove that so it can lock objects, generate and cache
+a DER encoding,, and so on.
+
+Another instance of thread-safety is when an object's internal state,
+such as cached values, are done with locks.
+One example of this is the reference counting API's described above.
+
+In all cases, however, it is generally not safe for one thread to
+mutate an object, such as setting elements of a private or public key,
+while another thread is using that object, such as verifying a signature.
+
+The same API's can often be used simultaneously on different objects
+without interference.
+For example, two threads can perform a signature using two different
+B<EVP_PKEY> objects.
+
+For implicit global state or singletons, thread-safety depends on the facility.
+The L<CRYPTO_secure_malloc(3)> and related API's have their own lock,
+while L<CRYPTO_malloc(3)> assumes the underlying platform allocation
+will do any necessary locking.
+Some API's, such as L<NCONF_load(3)> and related, or L<OBJ_create(3)>
+do no locking at all; this can be considered a bug.
+
+A separate, although related, issue is modifying "factory" objects
+when other objects have been created from that.
+For example, an B<SSL_CTX> object created by L<SSL_CTX_new(3)> is used
+to create per-connection B<SSL> objects by calling L<SSL_new(3)>.
+This is not safe, no matter if multiple threads are involved or not.
```

Comment:
I found this confusing. I think you're trying to say that modifying factory objects after objects have already been created is unsafe. But it reads like you are saying creating per-connection SSL objects from an SSL_CTX is unsafe.

Reply:
Yes, "modifying factory objects" is the concept I was looking for, thanks.